### ISSUE TITLE
Document adding extra modules

### ISSUE THREAD

**Issue by epelc** [openresty/docker-openresty#18]
It would be nice if there was some documentation on the recommend way to add extra modules(ie HeadersMore).

Is there a way to do it just using `FROM openresty/openrest:tag` or are we supposed to manually edit a dockerfile to add options under `./config`?

**Comment by neomantra** [comment]
Most of the README is about custom builds.  I do see that there's a broken link to a section in the top though.  If there's something you didn't get about it, please paste some suggestions here.

Either way, you are going to have to build an image yourself.   Since HeadersMore is included in the OpenResty distribution, you fortunately don't need to download it.  So the first way is best choice for that situation:

 * Clone [docker-openresty](https://github.com/openresty/docker-openresty) and use custom `RESTY_CONFIG_OPTIONS` as [described here](https://github.com/openresty/docker-openresty#building-non-rpm-based).  Note you don't need to edit the Dockerfile, use  `--build-arg`.

 * Create a Dockerfile with `FROM openresty/some-tag`.  Then download, build, and install dynamic modules and install whatever LuaRocks (including their dependencies).

If HeadersMore is a popular, secure module and not too big, we can include it in the default `RESTY_CONFIG_OPTIONS`.

**Comment by agentzh** [comment]
@neomantra I think ngx_headers_more should be enabled by default, just like the official binary packages provided by OpenResty:

http://openresty.org/en/linux-packages.html

**Comment by agentzh** [comment]
@neomantra BTW, the official OpenResty deb packages (for Ubuntu and Debian) will be out pretty soon. Hopefully we can migrate to those official deb packages in the docker images here after that ;)

**Comment by epelc** [comment]
I agree with enabling it by default. As a new user I suspected it would be included since this appears to be an official repo from OpenResty.
>Also, this module is included and enabled by default in the OpenResty bundle.
https://github.com/openresty/headers-more-nginx-module#installation

@neomantra Thanks for the info. I missed `RESTY_CONFIG_OPTIONS` when I was looking through the docs. Do you know what dir the modules are in so I can pass in `--add-module=/path/to/headers-more-nginx-module`? I'm sure I can find it if not but this might be good info for the readme too.

**Comment by epelc** [comment]
@neomantra @agentzh  turns out `headers-more` is already included.

You can run openresty with the `-V` flag to see all included modules/build options.
Example output for `openresty/openrest:xenial`
```
nginx version: openresty/1.11.2.1
built by gcc 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.2) 
built with OpenSSL 1.0.2h  3 May 2016
TLS SNI support enabled
configure arguments: --prefix=/usr/local/openresty/nginx --with-cc-opt=-O2 --add-module=../ngx_devel_kit-0.3.0 --add-module=../echo-nginx-module-0.60 --add-module=../xss-nginx-module-0.05 --add-module=../ngx_coolkit-0.2rc3 --add-module=../set-misc-nginx-module-0.31 --add-module=../form-input-nginx-module-0.12 --add-module=../encrypted-session-nginx-module-0.06 --add-module=../srcache-nginx-module-0.31 --add-module=../ngx_lua-0.10.6 --add-module=../ngx_lua_upstream-0.06 --add-module=../headers-more-nginx-module-0.31 --add-module=../array-var-nginx-module-0.05 --add-module=../memc-nginx-module-0.17 --add-module=../redis2-nginx-module-0.13 --add-module=../redis-nginx-module-0.3.7 --add-module=../rds-json-nginx-module-0.14 --add-module=../rds-csv-nginx-module-0.07 --with-ld-opt=-Wl,-rpath,/usr/local/openresty/luajit/lib --with-openssl=/tmp/openssl-1.0.2h --with-pcre=/tmp/pcre-8.39 --with-file-aio --with-http_addition_module --with-http_auth_request_module --with-http_dav_module --with-http_flv_module --with-http_geoip_module=dynamic --with-http_gunzip_module --with-http_gzip_static_module --with-http_image_filter_module=dynamic --with-http_mp4_module --with-http_random_index_module --with-http_realip_module --with-http_secure_link_module --with-http_slice_module --with-http_ssl_module --with-http_stub_status_module --with-http_sub_module --with-http_v2_module --with-http_xslt_module=dynamic --with-ipv6 --with-mail --with-mail_ssl_module --with-md5-asm --with-pcre-jit --with-sha1-asm --with-stream --with-stream_ssl_module --with-threads
```
specifically `--add-module=../headers-more-nginx-module-0.31`

Anyways I figured that out after getting duplicate module related errors when I got the build to work.

You can add any of the included modules by using `--add-module=../{MODULE_NAME}` for your `RESTY_CONFIG_OPTIONS`. The default module names available are in the `bundle` folder found in this tar archive https://openresty.org/download/openresty-1.11.2.1.tar.gz

**Comment by serain** [comment]
@epelc would it be possible to expand your answer with an example?

FWIW I'm not an openresty or even nginx user, I just needed to set something up quickly with `headers-more`. It's confusing that it's not enabled by default (as it apparently is supposed to be in openrusty?) and the docs do not help a newcommer understand how to enable a specific module.

**Comment by epelc** [comment]
@serain looks like you aren’t specifiying the openresty image and it is defaulting to nginx because of your service name. Lookup the image name option for docker compose and put in the openresty image name. It has more headers by default so you don’t need to do anything special. Same situation as me. I’m on mobile now so you’ll have to do the looking.

**Comment by neomantra** [comment]
You want to use `RESTY_CONFIG_OPTIONS_MORE` instead of `RESTY_CONFIG_OPTIONS`.  The `_MORE` one will *add* to the existing list rather than replacing the list.

so something like this in your section:
```
        RESTY_CONFIG_OPTIONS_MORE: --add-module=../headers-more-nginx-module-0.33
```
But yes, you also need a base image to build from and you aren't specifying it there.
